lib/Mojo/WebSocketProxy/Backend/JobAsync.pm

```diff
@@ -0,0 +1,101 @@
+package Mojo::WebSocketProxy::Backend::JobAsync;
+
+use strict;
+use warnings;
+
+use parent qw(Mojo::WebSocketProxy::Backend);
+
+no indirect;
+
+use IO::Async::Loop::Mojo;
+use Job::Async;
+use Scope::Guard qw(guard);
+
+use Log::Any qw($log);
+
+## VERSION
+
+sub new {
+    my ($class, %args) = @_;
+    # Avoid holding these - we only want the Job::Async::Client instance, and everything else
+    # should be attached to the loop (which sticks around longer than we expect to).
+    my $loop = delete $args{loop};
+    my $jobman = delete $args{jobman};
+
+    my $self = bless \%args, $class;
+
+    # We'd like to provide some flexibility for people trying to integrate this into
+    # other systems, so any combination of Job::Async::Client, Job::Async and/or IO::Async::Loop
+    # instance can be provided here.
+    $self->{client} //= do {
+        $jobman //= do {
+            # We don't hold a ref to this, since that might introduce unfortunate cycles
+            $loop //= do {
+                require IO::Async::Loop::Mojo;
+                IO::Async::Loop::Mojo->new;
+            };
+            $loop->add(
+                my $jobman = Job::Async->new
+            );
+            $jobman
+        };
+        $jobman->client;
+    };
+    return $self;
+}
+
+sub client { shift->{client} }
+
+sub call_rpc {
+    my ($self, $c, $req_storage) = @_;
+    my $method   = $req_storage->{method};
+    my $msg_type = $req_storage->{msg_type} ||= $req_storage->{method};
+
+    $req_storage->{call_params} ||= {};
+
+    my $rpc_response_cb = $self->get_rpc_response_cb($c, $req_storage);
+
+    my $before_get_rpc_response_hook = delete($req_storage->{before_get_rpc_response}) || [];
+    my $after_got_rpc_response_hook  = delete($req_storage->{after_got_rpc_response})  || [];
+    my $before_call_hook             = delete($req_storage->{before_call})             || [];
+    my $params = $self->make_call_params($c, $req_storage);
+    $log->debugf("method %s has params = %s", $method, $params);
+
+    $_->($c, $req_storage) for @$before_call_hook;
+    $self->client->submit(
+        data => $req_storage
+    )->then(sub {
+        my ($result) = @_;
+        $log->debugf('->submit completion: ', $result);
+
+        $_->($c, $req_storage) for @$before_get_rpc_response_hook;
+
+        # unconditionally stop any further processing if client is already disconnected
+        return Future->done unless $c and $c->tx;
+
+        my $mem_guard = guard {
```

Comment:
I am slightly dubious about the idea of using `Scope::Guard` in a Future chain. In this particular code it looks fine, but adding a second `->then` chain afterwards whose scope is supposed to contain the guard as well, would break it due to pad non-sharing of the `$mem_guard` lexical. I wonder if

    ->followed_by( sub { undef $req_storage; } )

would be better?
(Additionally I consider making `->finally` a name alias to `followed_by`)

Reply:
I think we should be able to tack an `->on_ready()` clause onto the end of the code, yes - there doesn't seem much benefit in `scope_guard` when we have timeout-controller Futures which are guaranteed to leave the `pending` state at some point.